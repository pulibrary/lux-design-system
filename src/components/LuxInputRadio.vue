<template>
  <component :is="wrapper" class="lux-input">
    <legend v-if="groupLabel">{{ groupLabel }}</legend>
    <div
      v-for="option in options"
      class="lux-radio"
      :class="{ 'lux-inline': !vertical }"
      :key="option.id"
    >
      <input
        type="radio"
        :id="option.id"
        :name="option.name"
        :value="option.value"
        :checked="option.checked"
        :disabled="option.disabled"
        :required="option.required"
        @change="change($event.target.value)"
        @blur="inputblur($event.target)"
      />
      <label v-if="option.label" :for="option.id">{{ option.label }}</label>
      <label v-else :for="option.id">{{ option.value }}</label>
    </div>
    <div role="alert" class="lux-error" v-if="errormessage">{{ errormessage }}</div>
  </component>
</template>

<script>
/**
 * Radio buttons should only be used when a user can select one option.
 * For multiple selections, use checkboxes.
 */
export default {
  name: "LuxInputRadio",
  status: "ready",
  release: "1.0.0",
  type: "Element",
  model: {
    prop: "checked",
    event: "change",
  },
  emits: ["change", "inputblur"],
  data: function () {
    return {
      wrapper: this.groupLabel.length ? "fieldset" : "div",
    }
  },
  props: {
    /**
     * If true, the radio buttons will be stacked vertically. Otherwise they will be horizontal (inline).
     */
    vertical: {
      type: Boolean,
      default: false,
    },
    /**
     * The available options to check. Option properties are: id, value, disabled, required, checked
     */
    options: {
      required: true,
      type: Array,
    },
    /**
     * The label of the form input field.
     */
    groupLabel: {
      type: String,
      default: "",
    },
    /**
     * The validation message a user should get.
     */
    errormessage: {
      type: String,
      default: "",
    },
    /**
     * Unique identifier of the form input field.
     */
    id: {
      type: String,
      default: "",
      required: true,
    },
    /**
     * Whether the form input field is disabled or not.
     * `true, false`
     */
    disabled: {
      type: Boolean,
      default: false,
    },
    /**
     * Whether the form input field is required or not.
     * `true, false`
     */
    required: {
      type: Boolean,
      default: false,
    },
    /**
     * Manually trigger input field’s hover state.
     * `true, false`
     */
    hover: {
      type: Boolean,
      default: false,
    },
    /**
     * Manually trigger input field’s focus state.
     * `true, false`
     */
    focus: {
      type: Boolean,
      default: false,
    },
  },
  methods: {
    change(value) {
      this.$emit("change", value)
    },
    inputblur(value) {
      this.$emit("inputblur", value)
    },
  },
}
</script>

<style lang="scss" scoped>
@import "../assets/styles/spacing.scss";
@import "../assets/styles/mixins.scss";
@import "../assets/styles/variables.css";
@import "../assets/styles/system.scss";
@import "../assets/styles/focus.scss";

fieldset {
  border: 0;
  padding: 0;
}

.lux-input {
  font-weight: var(--font-weight-regular);
  font-family: var(--font-family-text);
  font-size: var(--font-size-base);

  legend {
    margin-bottom: var(--space-x-small);
  }

  label {
    padding-right: var(--space-base);
  }
}

.lux-radio {
  font-family: var(--font-family-text);
  line-height: var(--line-height-base);
  display: flex;
}

.lux-radio label {
  cursor: pointer;
  padding-left: var(--space-xx-small);
  margin-top: calc(var(--font-size-base) * 0.175);
  margin-bottom: calc(var(--font-size-base) * 0.175);
}

.lux-inline {
  display: inline-block;
}

[disabled] + label {
  cursor: not-allowed;
  color: var(--color-grayscale);
}

input[type="radio"] {
  width: calc(var(--font-size-base) * 1.35);
  height: calc(var(--font-size-base) * 1.35);
}

input[type="radio"]:focus-visible {
  @include princeton-focus(dark);
}
</style>

<docs>
  ```jsx
  <div>
    <lux-input-radio
      id="foo"
      vertical groupLabel="Where is my mind?"
      :options="[
        {name: 'radio-group-name', value: 'In the clouds', id: 'radio-opt1', required: true},
        {name: 'radio-group-name', value: 'I don\'t know', id: 'radio-opt2', disabled: true}
      ]">
    </lux-input-radio>
  </div>
  ```
</docs>
