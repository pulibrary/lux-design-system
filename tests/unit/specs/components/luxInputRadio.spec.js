import { createLocalVue, mount } from "@vue/test-utils"
import { nextTick } from "vue"
import LuxInputRadio from "@/components/LuxInputRadio.vue"

describe("LuxInputRadio.vue", () => {
  let wrapper

  beforeEach(() => {
    wrapper = mount(LuxInputRadio, {
      slots: {
        default: "Apply Changes.",
      },
      propsData: {
        id: "foo",
        options: [
          {
            name: "opt 1",
            value: "one",
            id: "radio-opt1",
            checked: true,
          },
          {
            name: "opt 2",
            value: "two",
            id: "radio-opt2",
            checked: false,
          },
        ],
      },
    })
  })

  it("should emit a change event with the correct value when clicked", () => {
    wrapper.findAll("input").at(1).trigger("change")
    expect(wrapper.emitted().change[0]).toEqual(["two"])
  })

  it("should emit an inputblur event with the correct value when focus is blurred", () => {
    const radio = wrapper.findAll("input").at(1)
    radio.trigger("focus")
    radio.trigger("blur")
    const emitted = wrapper.emitted()
    expect(Object.prototype.hasOwnProperty.call(emitted, "inputblur")).toBe(true)
  })

  it("should have a label for the correct id", () => {
    const label = wrapper.findAll("label").at(1)
    expect(label.text()).toBe("two")
    expect(label.attributes().for).toBe("radio-opt2")
  })

  it("should stack if vertical is true", async () => {
    const radio = wrapper.findAll(".lux-radio").at(0)
    expect(radio.classes()).toContain("lux-inline")
    await wrapper.setProps({ vertical: true })
    expect(radio.classes()).not.toContain("lux-inline")
  })

  it("should display an errormessage with the proper role when passed in as a prop", async () => {
    expect(wrapper.find(".lux-error").exists()).toBe(false)
    await wrapper.setProps({ errormessage: "Something went wrong." })
    expect(wrapper.find(".lux-error").exists()).toBe(true)
    expect(wrapper.find(".lux-error").attributes().role).toBe("alert")
  })

  it("should display a legend if groupLabel is passed in as a prop", async () => {
    expect(wrapper.find("legend").exists()).toBe(false)
    await wrapper.setProps({ groupLabel: "Multiple choice:" })
    expect(wrapper.find("legend").exists()).toBe(true)
  })

  it("should have the id that is passed in as a prop", () => {
    const radio = wrapper.findAll("input").at(1)
    expect(radio.attributes().id).toBe("radio-opt2")
  })

  it("should have a fieldset wrapper if groupLabel is passed in as a prop", async () => {
    const fieldset = wrapper.find("fieldset")
    expect(fieldset.exists()).toBe(false)
    // can't setProps here because the component must be regenerated to change its wrapper
    const wrapper2 = mount(LuxInputRadio, {
      slots: {
        default: "Apply Changes.",
      },
      propsData: {
        id: "bar",
        groupLabel: "Multiple choice:",
        options: [
          {
            name: "opt 1",
            value: "one",
            id: "radio-opt1",
            checked: true,
          },
        ],
      },
    })
    const fieldset2 = wrapper2.find("fieldset")
    expect(fieldset2.exists()).toBe(true)
  })

  // checked and disabled are not showing up in snapshots
  // cannot seem to test
  //
  // it("should not be clickable if disabled", () => {
  //   wrapper.setProps({ options: [
  //     {
  //       name: "opt 1",
  //       value: "one",
  //       id: "radio-opt1",
  //       disabled: true,
  //       checked: false
  //     }
  //   ] })
  //   const radio = wrapper.findAll(".radio").at(0)
  //   radio.trigger("click")
  //   expect(radio.attributes().checked).toBe(false)
  // })

  it("has the expected html structure", () => {
    expect(wrapper.element).toMatchSnapshot()
  })
})
