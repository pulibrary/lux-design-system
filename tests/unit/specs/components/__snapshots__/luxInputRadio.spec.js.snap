// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LuxInputRadio.vue has the expected html structure 1`] = `
<div
  class="lux-input"
>
  <!--v-if-->
  
  <div
    class="lux-radio lux-inline"
  >
    <input
      id="radio-opt1"
      name="opt 1"
      type="radio"
      value="one"
    />
    <label
      for="radio-opt1"
    >
      one
    </label>
  </div>
  <div
    class="lux-radio lux-inline"
  >
    <input
      id="radio-opt2"
      name="opt 2"
      type="radio"
      value="two"
    />
    <label
      for="radio-opt2"
    >
      two
    </label>
  </div>
  
  <!--v-if-->
</div>
`;
